### ISSUE TITLE
[Python] How to do arrow table group by and split?

### ISSUE THREAD

**Issue by code1704** [apache/arrow#14882]
### Describe the usage question you have. Please include as many useful details as  possible.

How to group arrow table items and split into tables?

```

g = table.group_by("a")
for x in g:
   do_somthing(x)

```

### Component(s)

Python

**Comment by AlenkaF** [comment]
You can access the tables derived from a grouping by using a subsequent `aggregate()` method, see:

- https://arrow.apache.org/docs/dev/python/generated/pyarrow.Table.html#pyarrow.Table.group_by
- https://arrow.apache.org/cookbook/py/data.html#group-a-table
- https://arrow.apache.org/docs/python/compute.html#grouped-aggregations

**Comment by westonpace** [comment]
Aggregate methods, in particular the `list` aggregate method, can get you there but it is a bit clumsy I think:

```
import pyarrow as pa

tab = pa.Table.from_pydict({"key": [1, 1, 1, 2, 2], "value1": [1, 2, 3, 4, 5], "value2": ["a", "b", "c", "d", "e"]})
lists = tab.group_by("key").aggregate([("value1", "list"), ("value2", "list")])

keys = lists.column("key")
for i in range(lists.num_rows):
    print(f"Table for group: {keys[i]}")
    val1_col = lists.column("value1_list")[i].values
    val2_col = lists.column("value2_list")[i].values
    sub_table = pa.Table.from_arrays([val1_col, val2_col], names=["value1", "value2"])
    print(sub_table)
    print()
```

I would support a new API for this.  However, this will probably need changes at the C++ level.  An intuitive implementation I think would lead to a single exec plan that had multiple sinks (or we could just canonicalize the above usage of the list aggregate function and provide convenience for it).

This will be very useful, for example, for someone that wanted to partition a table based on some column (or the hash of a column) and distribute the work.

**Comment by yalwan-sage** [comment]
> This will be very useful, for example, for someone that wanted to partition a table based on some column (or the hash of a column) and distribute the work.

+1 for this, currently needing to go to/from pandas is not an exciting prospect

**Comment by ljishen** [comment]
@westonpace Can we allow the struct type of data to be the target of an aggregate? In this way, if we want to collect multiple columns in the output, we don't need to do like
```
lists = tab.group_by("key").aggregate([("value1", "list"), ("value2", "list")])
```
but maybe
```
lists = tab.group_by("key").aggregate([(["value1", "value2"], "list")])
```
Also, it may reduce the number of times to run the "hash_list" function from 2x to 1x in this case.

**Comment by westonpace** [comment]
> Can we allow the struct type of data to be the target of an aggregate?

No,and it probably wouldn't matter much for runtime.  Either way you need to aggregate all the data, group the keys, and then issue a `Take` for each array.

**Comment by ljishen** [comment]
Oh, you are right. Thanks @westonpace .

**Comment by JacekPliszka** [comment]
+1 for this

For me even much simpler thing: #37959  would work fine.

**Comment by willshiao** [comment]
Also ran into this issue, and totally agree that it would be an useful addition to PyArrow. But in the meantime, a workaround is to convert it to a Polars dataframe (see [`polars.from_arrow`](https://pola-rs.github.io/polars/py-polars/html/reference/api/polars.from_arrow.html)), which is mostly zero-copy. You can then use the [`partition_by`](https://pola-rs.github.io/polars/py-polars/html/reference/dataframe/api/polars.DataFrame.partition_by.html) function. It's much faster than doing the equivalent operation in Pandas.

**Comment by richardliaw** [comment]
would love to see a functionality like this canonicalized - is the groupby.aggregate(list) still the recommended path?

**Comment by e10v** [comment]
Here is the alternative proposed by ChatGPT :)


```python
import pyarrow as pa
import pyarrow.compute as pc

# Example data
table = pa.Table.from_pydict({
    "column_a": pa.array([1, 2, 1, 3, 2]),
    "column_b": pa.array(["x", "y", "z", "x", "y"])
})

group_col = "column_a"

# 1. Sort indices by the grouping column
sort_indices = pc.sort_indices(table, sort_keys=[(group_col, "ascending")])
sorted_table = table.take(sort_indices)

# 2. Identify boundaries where column_a changes
col = sorted_table[group_col]
# Compare adjacent values to find boundaries:
equal_to_previous = pc.equal(col[1:], col[:-1])
changes = pc.invert(equal_to_previous)  # True where value changes

# Convert boolean array to offsets
# Start at 0
offsets = [0]
# Add 1 to shift indices so they match starts of groups
for i, changed in enumerate(changes.to_pylist(), start=1):
    if changed:
        offsets.append(i)
# Final boundary is the length of the table
offsets.append(len(col))

# 3. Slice groups
for start, end in zip(offsets, offsets[1:]):
    subtable = sorted_table.slice(start, end - start)
    key = subtable[group_col][0].as_py()
    ...
```
